beacon-chain/db/filesystem/blob.go

```diff
@@ -323,11 +325,15 @@ func (bs *BlobStorage) shouldPrune(slot primitives.Slot) bool {
 
 // pruneOlderThan prunes blobs in the base directory based on the retention epoch and current slot.
 func (bs *BlobStorage) pruneOlderThan(slot primitives.Slot) error {
-	err := bs.Prune(slot)
-	if err != nil {
-		return err
+	v := bs.atomicPrunedEpoch.CompareAndSwap(uint64(bs.lastPrunedEpoch), uint64(slots.ToEpoch(slot)))
+	if v {
+		defer func() {
+			bs.lastPrunedEpoch = slots.ToEpoch(slot)
```

Comment:
This is mixing atomic and non-atomic values in a way that makes it tricky to reason about thread safety, and requires managing this extra value. I'm wondering if it would be cleaner using [`Uint64.Swap`](https://pkg.go.dev/sync/atomic#Uint64.Swap) instead. I also think you could get rid of `shouldPrune` if we decide to stop trying to avoid pruning at the epoch boundary, because the following code deals with the case where we've already triggered pruning for the given epoch. Strawman impl that I think makes some things simpler and reduces test surface area quite a bit:
```
func pruneBefore(latest primitives.Slot, offset primitives.Slot) primitives.Slot {
  // Safely compute the first slot in the epoch for the latest slot
  latest = latest - latest % params.BeaconConfig().SlotsPerEpoch
  if latest < offset {
    return 0
  }
  return latest - offset
}

func (bs *BlobStorage) tryPrune(latest primitives.Slot) {
  // This is assuming we compute bs.retentionSlots when the BlobStorage value is initialized.
  // This is a helpful simplifying assumption because it ensures Save never has to deal with an error from pruning.
  pruneBefore := pruneBefore(latest, bs.retentionSlots)
  // If we get the same value back from Swap, our job is done because it was already set and Prune was spawned.
  if bs.prunedBefore.Swap(pruneBefore) == pruneBefore {
    return
  }
  go func() { 
    // Rather than passing the current slot to Prune, pass it the slot where pruning should begin.
    // This simplifies the code and makes testing easier.
    if err := bs.Prune(pruneBefore); err != nil {
      log.WithError(err).Errorf("failed to prune blobs from slot %d", sidecar.Slot())
    }
  }
}
```

Just call `tryPrune` where you call `shouldPrune` now and I think bases are covered.

I snuck a couple other suggestions in here:
- If you compute the epoch where pruning should stop and use that as the atomic value, you can also pass it to `Prune` rather than the current slot. Now you only need to compute in one place - less complexity and tests needed for `Prune` and `processFolder`.
- Add `bufferEpochs` to `retentionEpochs` and convert to a `primitives.Slot` in `NewBlobStorage` and keep it around as `bs.retentionSlots`. It makes sense to validate this configurable is in bounds in the constructor, and it lets you avoid overflow error handling when working with slots for the rest of the code.
- Put all this together and we avoid pushing any error handling into Save, which shouldn't really care about Pruning errors as this is meant to be a "background" process.